### PR TITLE
ENH: PyArray_FromInterface checks descr if typestr is np.void

### LINKED ISSUE #5081: BUG np.broadcast_arrays does not work on void arrays with object fields
In the process of working on #4622, it was realised that `np.broadcas_arrays` do not work for void arrays with object fields, because in the process of changing new strides and shapes in `as_strided`, the `dtype` gets lost, and this cannot just be reset when object fields are present. While obviously this is somewhat of a corner case, it might still be good to see if it can be fixed easily, perhaps by doing something that preserves the dtype, or by bypassing the test for object fields.
```
import numpy as np
dt = np.dtype([('x', 'i8'), ('y', '?'), ('z', 'O')])
b = np.array([(1, True, None), (2, False, [3, 4, 5])], dtype=dt)
c = np.array([[-1], [-2]])
b_strided, c_strided = np.broadcast_arrays(b, c)
TypeError                                 Traceback (most recent call last)
...
/usr/lib/python3/dist-packages/numpy/lib/stride_tricks.py in as_strided(x, shape, strides)
     31     # Make sure dtype is correct in case of custom dtype
     32     if array.dtype.kind == 'V':
---> 33         array.dtype = x.dtype
     34     return array
     35 

TypeError: Cannot change data-type for object array.
```

### PR DESCRIPTION
**Pull request by jaimefrio**
When the 'typestr' member of the **array_interface** dictionary defines
a np.void dtype, check the 'descr' member, and if it is a valid dtype
description and it is not the default one, use it to construct the
dtype for the array to return.

This fixes #5081, as as_strided no longer has to worry about changing
the dtype of the return.

### CODE REVIEW

**Comment by jaimefrio** [comment]
There is still a hole in the ability of numpy to fully reconstruct an array, including its dtype, using `__array_interface__`. Right now any structured dtype is converted by `PyArray_FromInterface` into an array of type `np.void`. This PR fixes it for almost all structured dtypes except for one: a structured dtype of a single field, with name the empty string, and dtype `np.void`. This is hard to produce, as this will not give you that:

```
>>> dt = np.dtype([('', 'V8')])
dtype([('f0', 'V8')])
```

but if you really want it, you can get it:

```
>>> np.dtype({'names': [''], 'formats': ['V8']})
dtype([('', 'V8')])
```

**Comment by jaimefrio** [comment]
Added a little logic to `as_strided` so it can properly handle even that weird corner case.

File: numpy/core/src/multiarray/ctors.c [hunk 1]
```diff
@@ -2069,6 +2069,50 @@ PyArray_FromStructInterface(PyObject *input)
...
+    }
+    name = PyTuple_GET_ITEM(tuple, 0);
+    if (!(PyUString_Check(name) && PyUString_GET_SIZE(name) == 0)) {
+        return 0;
+    }
+    typestr2 = PyTuple_GET_ITEM(tuple, 1);
+#if defined(NPY_PY3K)
+    /* Allow unicode type strings */
+    if (PyUnicode_Check(typestr2)) {
+        tmp = PyUnicode_AsASCIIString(typestr2);
```
Comment (charris) [inline numpy/core/src/multiarray/ctors.c]: Looks like tmp can be NULL, assuming that the string cannot be converted to ascii. I'm not sure what the policy for this codec is.

File: numpy/core/src/multiarray/ctors.c [hunk 2]
```diff
@@ -2069,6 +2069,50 @@ PyArray_FromStructInterface(PyObject *input)
...
+        typestr2 = tmp;
+    }
+#endif
+    if (PyBytes_Check(typestr2) &&
+            PyObject_RichCompareBool(typestr, typestr2, Py_EQ)) {
+        ret = 1;
+    }
+#if defined(NPY_PY3K)
+    if (tmp == typestr2) {
+        Py_DECREF(tmp);
```
Comment (charris) [inline numpy/core/src/multiarray/ctors.c]: Might want `Py_XDECREF` if tmp is NULL. Not sure how to handle the error case if one is possible.

**Comment by charris** [comment]
Looks like there may be some conflicts with #5371, so lets get that in first. You are welcome to merge it :0

**Comment by charris** [comment]
@jaimefrio  Needs rebase.

**Comment by jaimefrio** [comment]
Rebased it is. Still need to figure how to handle that potential failure in the ASCII conversion...

**Comment by jaimefrio** [comment]
I think I have now taken care of the possible error: if we cannot convert to ASCII, then it isn't the default `descr`, so we return 0. `tmp` is set to NULL at the beginning, and `Py_XDECREF`'ed at the end just to be sure.

File: numpy/core/src/multiarray/ctors.c [hunk 3]
```diff
@@ -2069,6 +2069,51 @@ PyArray_FromStructInterface(PyObject *input)
...
+#if defined(NPY_PY3K)
+    PyObject *tmp = NULL;
+#endif
+    int ret = 0;
+
+    if (!PyList_Check(descr) || PyList_GET_SIZE(descr) != 1) {
+        return 0;
+    }
+    tuple = PyList_GET_ITEM(descr, 0);
+    if (!(PyTuple_Check(tuple) && PyTuple_GET_SIZE(tuple) == 2)) {
```
Comment (charris) [inline numpy/core/src/multiarray/ctors.c]: I like the way this condition is expressed better than the one above ;)

**Comment by charris** [comment]
LGTM, thanks Jaime.
